paper/paper.md

```diff
@@ -0,0 +1,109 @@
+
+### Heuristic Implementation Example
+
+```kotlin
+override fun simulate(node: NodeType): Double {
+    /*... Original Simulation code ...*/
+    while(true) {
+        val validActions = mdp.actions(currentState)
+        var bestActionScore = Int.MIN_VALUE // Begin heuristic
+            var bestActions = mutableListOf<Point>()
+            for (action in validActions) {
+                val score = heuristicWeight[action.x][action.y]
+                if (score > bestActionScore) {
+                    bestActionScore = score
+                    bestActions = mutableListOf(action)
+                }
+                if (score == bestActionScore) {bestActions.add(action)}
+            }
+        val randomAction = bestActions.random() // End heuristic
+        val newState = mdp.transition(currentState, randomAction)
+        /*... Original Simulation code ...*/
+    }
+}
+```
+
+# Results
+
+When the MCTS solver is accurately selecting the optimal solutions, it will continue to compel the agent to explore in the optimal subset of the state space, and reduce its exploration in the non-optimal subset of the state space. We provide a quick example in the MDP Domain of GridWorld, detailed in [@Liu:2021]. The cumulative number of visits corresponding to the optimal policy is proportionally increasing with respect to the number of MCTS iterations. Whereas for non-optimal solutions, the cumulative visits are significantly less because the solver will avoid visiting the non-optimal state subspace. 
+
+![Convergence of visits](gw_visits.png?raw=true "Title")
+
+# Conclusion
+
+In closing, *mctreesearch4j* presents a framework which enables programmers to adapt an MCTS solver to a variety of MDP domains. This is important because software application was a main focus of *mctreesearch4j*. Furthermore, *mctreesearch4j* is fully compatible with JVM, and this design decision was made due to the excellent support of class structure and generic variable typing in Kotlin, and other JVM languages, as well as support for mobile applications. Yet most importantly, *mctreesearch4j* is modular and extensible, the key mechanism of MCTS are broken down, and the programmer is able inherit class characteristics, redefine and/or re-implement certain sections of the algorithm while maintaining a high degree of MCTS standardization.
```

Comment:
```suggestion
In closing, *mctreesearch4j* presents a framework which enables programmers to adapt an MCTS solver to a variety of MDP domains. This is important because software application was a main focus of *mctreesearch4j*. Furthermore, *mctreesearch4j* is fully compatible with JVM, and this design decision was made due to the excellent support of class structure and generic variable typing in Kotlin, and other JVM languages, as well as support for mobile applications. Yet most importantly, *mctreesearch4j* is modular and extensible, the key mechanism of MCTS are broken down, and the programmer is able inherit class functions, redefine and/or re-implement certain sections of the algorithm while maintaining a high degree of MCTS standardization.
```

Reply:
`class members`